### ISSUE TITLE
Parsing '[ (]([ (]([ (]([ (](...' takes quadratic time

### ISSUE THREAD

**Issue by mity** [commonmark/cmark#285]
```
$ time py -c 'print("[ (](" * 10000)' | ./cmark >/dev/null
real    0m0.232s
user    0m0.030s
sys     0m0.000s

$ time py -c 'print("[ (](" * 40000)' | ./cmark >/dev/null
real    0m2.168s
user    0m0.000s
sys     0m0.045s

$ time py -c 'print("[ (](" * 80000)' | ./cmark >/dev/null
real    0m8.278s
user    0m0.016s
sys     0m0.046s
```

(Originally reported by @andersk for MD4C in https://github.com/mity/md4c/issues/57 but applies to Cmark too.)

**Comment by mity** [comment]
In MD4C, it was fixed by https://github.com/mity/md4c/commit/966b8e39b5721a3f881e27a2750c6e01a93f87c8.

See the specs:

> A link title consists of either
> * a sequence of zero or more characters between straight double-quote characters (`"`), including a `"` character only if it is backslash-escaped, or
> * a sequence of zero or more characters between straight single-quote characters (`'`), including a `'` character only if it is backslash-escaped, or
> * a sequence of zero or more characters between matching parentheses (`(...)`), including a `)` character only if it is backslash-escaped.

Imho the 3rd option has to be updated to disallow also `(`. Otherwise, given this test case, you have to scan till the end of the input for every part of the input pattern, leading to O(n^2). Furthermore allowing it leads to very unreadable and unintuitive behavior:

```
$ echo '[x](url (crazy_title())' | ./cmark
<p><a href="url" title="crazy_title(">x</a></p>
```

**Comment by mity** [comment]
This was also already touched here: https://github.com/commonmark/CommonMark/issues/526

**Comment by jgm** [comment]
Fixed by 9f7d0a674b0a9c741daf359516d80dbf13ec960a
